Part of Issue2099: Add fees to TxBuilder backwards compatible Faucet

hewison-chris opened the pull request:
To enable the Faucet to still use TxBuilder before being updated to use the new behavior the original `sign` can still be used.

hewison-chris committed:
Update TxBuilder to include fees

When a `transaction` is created then the sum of the input values minus
the sum of the output values is the implied `fee` for the transaction.
The `fee` must be at least the transaction size in bytes multiplied by
the minimum fee rate as defined in the `Consensus` parameters.

Geod24 commented:
> Unit test ValidatorRecurringEnrollment fix

What was wrong ? Please add it to the commit message.

kchulj commented:
What should be merged first, this or #2371 ?

kchulj commented:
Faucet and Agora work with this commit. (although there are other bugs that need to be fixed)

hewison-chris commented:
> What should be merged first, this or #2371 ?

#2371 should be first. It is almost ready.

hewison-chris committed:
Trivial remove of nested ternary

hewison-chris commented:
~~Still not quite ready yet. Will try and fix it over the weekend / holiday.~~

hewison-chris committed:
Include small change as extra fees

We only create refund outputs for usable `Amounts` which would be able
to cover typical fee cost. This small change is just included as extra
fee for the transaction.

kchulj commented:
1st and 2nd commit LG.

hewison-chris commented:
This is now passing ci as the deprecated `sign` is kept for the Faucet to use. The small change outputs are prevented by adding them to fees.

hewison-chris commented:
Closing as we moved the required changes to PR https://github.com/bosagora/agora/pull/2388
